### PR TITLE
SONARSLANG-22 Support the import of Android Lint issues reports

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

File: its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java [hunk 1]
```diff
@@ -54,6 +54,26 @@ public void detekt() {
     }
   }
 
+  @Test
+  public void android_lint() {
+    SonarScanner sonarScanner = getSonarScanner(BASE_DIRECTORY, "androidlint");
+    sonarScanner.setProperty("sonar.androidLint.reportPaths", "lint-results.xml");
+    ORCHESTRATOR.executeBuild(sonarScanner);
+    List<Issue> issues = getExternalIssues();
+    if (ORCHESTRATOR.getServer().version().isGreaterThanOrEquals(7, 2)) {
```
Comment (vilchik-elena) [inline its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java]: extract to variable?

File: its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java [hunk 2]
```diff
@@ -54,6 +54,26 @@ public void detekt() {
...
+      assertThat(issues).hasSize(1);
```
Comment (vilchik-elena) [inline its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java]: can we test other extension?

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+public class AndroidLintRulesDefinition implements RulesDefinition {
+
+  private static final String RULES_JSON = "org/sonar/l10n/android/rules/androidlint/rules.json";
+
+  /**
+   * Android lint scopes could be: ".xml", ".java", ".kt", ".kts", ".properties", ".gradle", "proguard.cfg", "proguard-project.txt", ".png", ".class"
+   * ( https://android.googlesource.com/platform/tools/base/+/studio-master-dev/lint/libs/lint-api/src/main/java/com/android/tools/lint/detector/api/Scope.kt )
+   * But this sensor provides rule descriptions only for ".xml", ".java", ".kt"
+   */
+  private static final List<String> LANGUAGES_WHO_BENEFITS_FROM_RULE_DEFINITIONS = Arrays.asList("xml", "java", SlangPlugin.KOTLIN_LANGUAGE_KEY);
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java]: `RULE_REPOSITORY_LANGUAGES`

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+  private static final List<String> TEXT_FILE_EXTENSIONS = Arrays.asList(".xml", ".java", ".kt", ".kts", ".properties", ".gradle", ".cfg", ".txt");
+
+  static final List<ExternalRuleLoader> RULE_LOADERS = LANGUAGES_WHO_BENEFITS_FROM_RULE_DEFINITIONS.stream()
+    .map(language -> new ExternalRuleLoader(LINTER_KEY + "-" + language, LINTER_NAME, RULES_JSON, language))
+    .collect(Collectors.toList());
+
+  private static final String DEFAULT_REPOSITORY_WITHOUT_RULE_DEFINITIONS = LINTER_KEY;
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java]: IMO `DEFAULT_REPOSITORY_KEY` is enough

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  static RuleKey ruleKey(@Nullable String language, String ruleId) {
+    if (language == null || !LANGUAGES_WHO_BENEFITS_FROM_RULE_DEFINITIONS.contains(language)) {
+      return RuleKey.of(DEFAULT_REPOSITORY_WITHOUT_RULE_DEFINITIONS, ruleId);
+    }
+    return RuleKey.of(LINTER_KEY + "-" + language, ruleId);
+  }
+
+  static boolean isTextFile(String file) {
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java]: may be this method should be moved to `AndroidLintXmlReportReader`?

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintXmlReportReader.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.issue.NewExternalIssue;
+import org.sonar.api.batch.sensor.issue.NewIssueLocation;
+import org.sonar.api.rule.RuleKey;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonarsource.analyzer.commons.ExternalRuleLoader;
+
+class AndroidLintXmlReportReader {
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintXmlReportReader.java]: can we find a better name? this class doesn't only read by also saves issues

File: its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java (hunk 1 above)
Comment (alban-auzeill) [reply its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java]: extracted as externalIssuesSupported

File: its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java (hunk 2 above)
Comment (alban-auzeill) [reply its/plugin/src/test/java/com/sonarsource/slang/ExternalReportTest.java]: I added a build.gradle file

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java (hunk 3 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java]: renamed

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java (hunk 4 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java]: renamed

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java (hunk 5 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRulesDefinition.java]: I want to keep RULE_REPOSITORY_LANGUAGES and TEXT_FILE_EXTENSIONS in the same file.

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintXmlReportReader.java (hunk 6 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintXmlReportReader.java]: I removed the save issue method from this class

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      LOG.info("Importing {}", reportPath);
+      AndroidLintXmlReportReader.read(in, (id, file, line, message) -> saveIssue(context, id, file, line, message));
+    } catch (IOException | XMLStreamException e) {
+      LOG.error("No issues information will be saved as the report file '{}' can't be read.", reportPath, e);
+    }
+  }
+
+  private static void saveIssue(SensorContext context, String id, String file, String line, String message) {
+    if (id.isEmpty() || message.isEmpty() || file.isEmpty() || !AndroidLintRulesDefinition.isTextFile(file)) {
+      return;
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java]: @alban-auzeill we need log here

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+  public void execute(SensorContext context) {
+    List<File> reportFiles = ExternalReportProvider.getReportFiles(context, REPORT_PROPERTY_KEY);
+    reportFiles.forEach(report -> importReport(report, context));
+  }
+
+  private static void importReport(File reportPath, SensorContext context) {
+    try (InputStream in = new FileInputStream(reportPath)) {
+      LOG.info("Importing {}", reportPath);
+      AndroidLintXmlReportReader.read(in, (id, file, line, message) -> saveIssue(context, id, file, line, message));
+    } catch (IOException | XMLStreamException e) {
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java]: add runtimeexception

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/SlangPlugin.java [hunk 9]
```diff
@@ -65,12 +71,20 @@ public void define(Context context) {
...
-          .category(KOTLIN_CATEGORY)
-          .subCategory(EXTERNAL_LINTER_SUBCATEGORY)
+          .category(EXTERNAL_ANALYZERS_CATEGORY)
+          .subCategory(KOTLIN_SUBCATEGORY)
+          .onQualifiers(Qualifiers.PROJECT)
+          .multiValues(true)
+          .build(),
+        PropertyDefinition.builder(AndroidLintSensor.REPORT_PROPERTY_KEY)
+          .name("Android Lint Report Files")
+          .description("Paths (absolute or relative) to xml files with android lint issues.")
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/SlangPlugin.java]: IMO should be **A**ndroid **L**int

File: utils/generate-android-lint-rules/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRuleDefinition.java [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+        rule.name = consumeHeader("Summary");
+        String priority = consumeHeader("Priority");
+        String severity = consumeHeader("Severity");
+        String category = consumeHeader("Category");
+        rule.description = consumeDescription();
+
+        rule.type = mapType(severity, category, priority);
+        rule.severity = mapSeverity(severity);
+        rule.tags = new HashSet<>(mapTags(category));
+        rule.tags.add("android");
```
Comment (vilchik-elena) [inline utils/generate-android-lint-rules/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRuleDefinition.java]: let's remove it

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java (hunk 7 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java]: added

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java (hunk 8 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java]: done for android lint and detekt

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/ExternalReportTestUtils.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+  public static SensorContextTester createContext(Path projectDir, int majorVersion, int minorVersion) throws IOException {
+    SensorContextTester context = SensorContextTester.create(projectDir);
+    Files.list(projectDir)
+      .forEach(file -> addFileToContext(context, projectDir, file));
+    context.setRuntime(SonarRuntimeImpl.forSonarQube(Version.create(majorVersion, minorVersion), SonarQubeSide.SERVER));
+    return context;
+  }
+
+  public static void assertLogsContainOnlyInfo(LogTester logTester) {
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/ExternalReportTestUtils.java]: name doesn't correspond to content of method

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/SlangPlugin.java (hunk 9 above)
Comment (vilchik-elena) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/SlangPlugin.java]: @alban-auzeill don't miss this one

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensorTest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+import static com.sonarsource.slang.externalreport.ExternalReportTestUtils.assertLogsContainOnlyInfo;
+import static com.sonarsource.slang.externalreport.ExternalReportTestUtils.createContext;
+import static com.sonarsource.slang.externalreport.ExternalReportTestUtils.executeSensor;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class AndroidLintSensorTest {
+
+  private static final Path PROJECT_DIR = Paths.get("src", "test", "resources", "externalreport", "androidlint");
+
+  private static AndroidLintSensor detektSensor = new AndroidLintSensor();
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensorTest.java]: rename variable

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensorTest.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    assertThat(externalIssues).isEmpty();
+    assertLogsContainOnlyInfo(logTester);
+  }
+
+  @Test
+  public void no_issues_with_invalid_report_path() throws IOException {
+    SensorContextTester context = createContext(PROJECT_DIR, 7, 2);
+    context.settings().setProperty("sonar.androidLint.reportPaths", resolveInProject("invalid-path.txt"));
+    List<ExternalIssue> externalIssues = executeSensor(detektSensor, context);
+    assertThat(externalIssues).isEmpty();
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensorTest.java]: these 3 assertions could be refactored

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+      LOG.error("No issues information will be saved as the report file '{}' can't be read.", reportPath, e);
+    }
+  }
+
+  private static void saveIssue(SensorContext context, String id, String file, String line, String message) {
+    if (id.isEmpty() || message.isEmpty() || file.isEmpty() || !AndroidLintRulesDefinition.isTextFile(file)) {
+      LOG.debug("Missing information or unsupported file type for:'{}', file:'{}', message:'{}'", id, file, message);
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java]: it's not clear from message what is first element (rule id?)

**Comment by vilchik-elena** [comment]
@alban-auzeill rename gif file

File: sonar-slang-plugin/src/test/resources/externalreport/androidlint/lint-results-with-errors.xml [hunk 15]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<issues format="4" by="lint 26.1.1">
+
+    <issue
+        id="UnknownRuleKey"
+        message="Unknown rule.">
+        <location file="AndroidManifest.xml"/>
```
Comment (vilchik-elena) [inline sonar-slang-plugin/src/test/resources/externalreport/androidlint/lint-results-with-errors.xml]: test multiple locations

File: utils/generate-android-lint-rules/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRuleDefinition.java (hunk 10 above)
Comment (alban-auzeill) [reply utils/generate-android-lint-rules/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintRuleDefinition.java]: done

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/SlangPlugin.java (hunk 9 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/SlangPlugin.java]: done

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/ExternalReportTestUtils.java (hunk 11 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/ExternalReportTestUtils.java]: renamed to assertNoErrorWarnDebugLogs

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensorTest.java (hunk 13 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/test/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensorTest.java]: refactored into "executeSensorImporting("lint-results.xml")"

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java (hunk 14 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/externalreport/androidlint/AndroidLintSensor.java]: added 'id:'

File: sonar-slang-plugin/src/test/resources/externalreport/androidlint/lint-results-with-errors.xml (hunk 15 above)
Comment (alban-auzeill) [reply sonar-slang-plugin/src/test/resources/externalreport/androidlint/lint-results-with-errors.xml]: done

**Pull request merged**